### DISCUSSION TITLE
Roadmap March 2023

### DISCUSSION THREAD

**Discussion by ocots** [control-toolbox/control-toolbox.github.io#18] [General]
Starting tasks: 

- [ ] @jbcaillau @joseph-gergaud @PierreMartinon [Box and pure constraints for the direct method](https://github.com/control-toolbox/CTBase.jl/issues/4).
- [ ] @jbcaillau @ocots [Getters of multipliers associated to labeled constraints](https://github.com/control-toolbox/OptimalControl.jl/issues/116).
- [ ] @ocots @joseph-gergaud @PierreMartinon [Functional: Bolza cost](https://github.com/control-toolbox/OptimalControl.jl/issues/82).
- [x] @joseph-gergaud @PierreMartinon [Direct solve by Crank-Nicolson](https://github.com/control-toolbox/OptimalControl.jl/issues/117).
- [ ] @joseph-gergaud @PierreMartinon [Direct solve by generic RK scheme](https://github.com/control-toolbox/OptimalControl.jl/issues/117).
- [ ] @joseph-gergaud @PierreMartinon [Sparse usage (Jacobian, Hessian of the constraints)](https://github.com/control-toolbox/OptimalControl.jl/issues/118).
- [x] @PierreMartinon [Julia wrapper of bocop for benchmarking](https://github.com/control-toolbox/OptimalControl.jl/issues/119).
- [x] @ocots @PierreMartinon [Common optimal control solution with plot for the direct methods](https://github.com/control-toolbox/OptimalControl.jl/issues/120).
- [ ] @ocots @PierreMartinon [Common optimal control solution with init for the direct methods](https://github.com/control-toolbox/OptimalControl.jl/issues/120).
- [x] @ocots [improve solution (control) flow plot](https://github.com/control-toolbox/OptimalControl.jl/issues/81).
- [ ] @ocots @joseph-gergaud [Ipopt with the direct shooting method](https://github.com/control-toolbox/OptimalControl.jl/issues/121).
- [ ] @ocots @PierreMartinon [Replace the print callback of ipopt](https://github.com/control-toolbox/OptimalControl.jl/issues/55).

You can get access to the [list of all the issues labeled `march2023`](https://github.com/issues?q=is%3Aopen+is%3Aissue+user%3Acontrol-toolbox+archived%3Afalse+label%3Amarch2023).

**Remarks / other possible tasks.**

- we have to start the documentation!
- we could add an indirect shooting method: see [Issue 24](https://github.com/control-toolbox/OptimalControl.jl/issues/24).
- for the direct shooting method, we could generalize the problem solved by the unconstrained solver (need to generalize also the adjoint method for better performance). See [Issue 1](https://github.com/control-toolbox/CTDirectShooting.jl/issues/1). We have to add the option of the state discretization: see also [Issue 2](https://github.com/control-toolbox/CTDirectShooting.jl/issues/2) and [Issue 9](https://github.com/control-toolbox/HamiltonianFlows.jl/issues/9).
- benchmark of the direct shooting method with and without adjoint method
- usage of `assert macro` for pre-conditions: see [Issue 74](https://github.com/control-toolbox/OptimalControl.jl/issues/74).
- [Dynamics: output and in-place](https://github.com/control-toolbox/OptimalControl.jl/issues/71)
- we need to add problems in `CTProblems` starting by adding the solution of the Goddard problem.

**Comment by ocots** [comment]
We must handle print of OptimalControlSolution from `CTBase.jl`:
```julia
> sol = solve(ocp)
Error showing value of type OptimalControlSolution:
ERROR: MethodError: no method matching get(::IOContext{Base.TTY}, ::Symbol, ::Bool)
You may have intended to import Base.get

Some of the types have been truncated in the stacktrace for improved reading. To emit complete information
in the stack trace, evaluate `TruncatedStacktraces.VERBOSE[] = true` and re-run the code.

Closest candidates are:
...
```

**Reply by ocots** [reply]
I have added for the moment a method to `Base.show` in `OptimalControl.jl` (that should be moved to `CTBase.jl`.
```julia
import Base: show, Base
# we get an error when a solution is printed so I add this function
# which has to be put in the package CTBase and has to be completed
function Base.show(io::IO, ::MIME"text/plain", sol::OptimalControlSolution)
    nothing
end
```

**Comment by ocots** [comment]
When testing `OptimaControl.jl` I have the following message:
```bash
WARNING: both CTBase and Base export "adjoint"; uses of it in module OptimalControl must be qualified
WARNING: both CTBase and Base export "success"; uses of it in module OptimalControl must be qualified
```
Maybe, in `CTBase.jl`, add
```julia
import Base: adjoint, success
```
to extend the functions.

**Reply by ocots** [reply]
I am not really sure that getters is a good thing. It is quite simple and nice to access directly the field of struct. In that case, there is no need of `adjoint` and `success` methods in `CTBase.jl`.

**Comment by ocots** [comment]
I have also this message:
```bash
WARNING: importing deprecated binding Colors.RGB1 into PlotUtils.
WARNING: importing deprecated binding Colors.RGB4 into PlotUtils.
```

**Comment by ocots** [comment]
About Typing, maybe I should remove the Function Type in:

* https://github.com/control-toolbox/CTBase.jl/blob/main/src/macros.jl#L15
* https://github.com/control-toolbox/CTBase.jl/blob/main/src/solutions.jl#L29

since for instance, when using `Interpolations.jl` the interpolation function is not a subtype of `Function` (which is sad).

**Comment by ocots** [comment]
Coding Jupyter notebook and github: https://docs.github.com/en/repositories/working-with-files/using-files/working-with-non-code-files#working-with-jupyter-notebook-files-on-github

**Reply by ocots** [reply]
We can enable rich diff for jupyter notebooks:
![Capture d’écran 2023-03-02 à 11 09 48](https://user-images.githubusercontent.com/66357348/222399000-1d0d945c-3dc9-4b1b-b3f3-aad128a47634.png)
![Capture d’écran 2023-03-02 à 11 09 38](https://user-images.githubusercontent.com/66357348/222399182-10912a81-6025-4b71-a504-4554a1836923.png)

**Reply by PierreMartinon** [reply]
Nice !

**Comment by ocots** [comment]
@jbcaillau I have made a new release of `OptimalControl` (`v0.3.0`). See the notebooks in the `examples` folder.

**Comment by ocots** [comment]
@PierreMartinon I have removed internal getters, that is getters of for instance the `DirectSolution` since this struct is internal to `CTDirect`. Defining getters produce more functions and we have to be careful since the package `Base` has already functions like `adjoint` or `success`.

**Reply by PierreMartinon** [reply]
Well if the getter was just returning the field, we can choose direct access yes. We can still reintroduce getters later if we decide so at some point.

**Comment by jbcaillau** [comment]
Issue in [box constraints](https://github.com/control-toolbox/CTBase.jl/issues/4) transferred to `CTBase.jl`. Some other issues should be moved... (will almost surely break some links, though)

**Reply by ocots** [reply]
I had put all the issues in `OptimalControl.jl` since the labels (or tags) are by repository.  But no problem, I will add a label `march2023` in `CTBase.jl`.

**Comment by ocots** [comment]
We have to be better on the covering, that is we need more unit tests.

<img width="1034" alt="Capture d’écran 2023-03-10 à 22 51 02" src="https://user-images.githubusercontent.com/66357348/224436019-0d8ab925-b4d3-4b78-86b9-468b1b25dd12.png">

**Comment by ocots** [comment]
I have replaced `CTProblemLibrary` by `CTProblems` and `HamitlonianFlows` by `CTFlows`.

Now, we have to update any package which need it, that is update the `Project.toml`.

```bash
julia> ]
rm CTProblemLibrary HamitlonianFlows
add CTProblems CTFlows
```
